File: jarboot-ui2/src/services/FileService.ts
import Request from '@/common/Request';
import type { FileNode } from '@/types';
import CommonUtils from '@/common/CommonUtils';

const urlBase = '/api/jarboot/file-manager';

/**
 * File service
 */
export default class FileService {
  /**
   * 获取版本
   * @param baseDir
   * @param withRoot
   * @returns {Promise<any>}
   */
  public static getFiles(baseDir: string, withRoot: boolean, clusterHost?: string): Promise<any> {
    const form = new FormData();
    form.append('baseDir', baseDir);
    if (clusterHost) {
      form.append('clusterHost', clusterHost);
    }
    form.append('withRoot', withRoot + '');
    return Request.post<FileNode[]>(`${urlBase}/list`, form);
  }

  /**
   * 获取文件内容
   * @param path
   */
  public static getContent(path: string, clusterHost?: string) {
    const form = new FormData();
    form.append('path', path);
    if (clusterHost) {
      form.append('clusterHost', clusterHost);
    }
    return Request.post<string>(`${urlBase}/file/text`, form);
  }

  /**
   * 解析路径
   * @param node
   * @param baseDir
   */
  public static parseFilePath(node: any, baseDir: string) {
    let path = node.data.name;
    let parent = node.parent;
    if (!node.data?.parent) {
      return baseDir;
    }
    while (parent?.data.parent) {
      path = parent.data.name + '/' + path;
      parent = parent.parent;
    }
    if (baseDir) {
      path = baseDir + '/' + path;
    }
    return path;
  }

  /**
   * 删除文件
   * @param path
   * @param clusterHost
   */
  public static deleteFile(path: string, clusterHost?: string) {
    return Request.delete<string>(`${urlBase}/file/delete?path=${path}&clusterHost=${clusterHost ?? ''}`, {});
  }

  /**
   * 写文件
   * @param path
   * @param content
   */
  public static writeFile(path: string, content: string, clusterHost?: string) {
    const form = new FormData();
    form.append('path', path);
    form.append('content', content);
    if (clusterHost) {
      form.append('clusterHost', clusterHost);
    }
    return Request.post<string>(`${urlBase}/text`, form);
  }

  /**
   * 新建文件
   * @param path
   * @param content
   */
  public static newFile(path: string, content: string, clusterHost?: string) {
    const form = new FormData();
    form.append('path', path);
    form.append('content', content);
    if (clusterHost) {
      form.append('clusterHost', clusterHost);
    }
    return Request.post<string>(`${urlBase}/text/create`, form);
  }

  /**
   * 新增文件夹
   * @param path
   */
  public static addDirectory(path: string, clusterHost?: string) {
    const form = new FormData();
    form.append('path', path);
    if (clusterHost) {
      form.append('clusterHost', clusterHost);
    }
    return Request.post<string>(`${urlBase}/directory`, form);
  }

  public static download(path: string, name: string, callback: (result: boolean, msg?: string) => void, clusterHost?: string) {
    const form = new FormData();
    form.append('path', path);
    if (clusterHost) {
      form.append('clusterHost', clusterHost);
    }
    CommonUtils.download(`${urlBase}/file/download`, name, 'POST', form, callback);
  }
}

File: jarboot-server/src/main/java/io/github/majianzheng/jarboot/service/FileService.java
package io.github.majianzheng.jarboot.service;

import io.github.majianzheng.jarboot.api.pojo.FileNode;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.List;

/**
 * 文件管理
 * @author mazheng
 */
public interface FileService {
    /**
     * 获取文件列表
     * @param baseDir 路径
     * @param withRoot 是否包含根路径
     * @return 文件节点
     */
    List<FileNode> getWorkspaceFiles(String baseDir, boolean withRoot);

    /**
     * 获取文件内容
     * @param file 文件相对于工作目录的路径
     * @return 文件内容
     */
    String getContent(String file);

    /**
     * 删除文件
     * @param file 文件相对于工作目录的路径
     */
    void deleteFile(String file);

    /**
     * 上传文件
     * @param file 文件相对于工作目录的路径
     * @param is 文件流
     */
    void uploadFile(String file, InputStream is);

    /**
     * 下载文件
     * @param file 文件相对于工作目录的路径
     * @param os 文件流
     */
    void download(String file, OutputStream os);

    /**
     * 修改文件
     * @param file 文件相对于工作目录的路径
     * @param content 文件内容
     * @return key
     */
    String writeFile(String file, String content);

    /**
     * 新建文件
     * @param file 文件相对于工作目录的路径
     * @param content 文件内容
     * @return key
     */
    String newFile(String file, String content);

    /**
     * 新建文件夹
     * @param file 文件相对于工作目录的路径
     * @return key
     */
    String addDirectory(String file);
}
